### PR TITLE
feat(embedded): embed a single chart as an independent entity [SIP-145]

### PR DESCRIPTION
**Pull request by amaannawab923**
### SUMMARY

https://github.com/user-attachments/assets/7db5bea2-34e3-4f85-a1be-d45a37cd387c

Adds the ability to embed a **single chart** as an independent entity, rather than only
whole dashboards. A chart gets its own embed UUID, its own guest token scoped to that
chart alone, and its own allowed-domain list, and is served through the existing
`/embedded/<uuid>` route.

This reimplements the approach from #33424 against current `master`. That PR has been
open since May 2025 and has drifted; this branch is a fresh implementation rather than a
rebase, and #33424 is left open and untouched for reference.

**Approach.** An embedded chart renders through the existing dashboard chart stack —
`gridComponents/Chart`, its header controls, drill and cross-filter plumbing — by
synthesising the minimum slice of dashboard state a single chart needs. That is what keeps
the header menu, drill, `View query`, `View as table`, exports and fullscreen working with
no reimplementation.

Notably this reuses `HYDRATE_DASHBOARD` rather than introducing a parallel
`HYDRATE_EMBEDDED` action, so **no dashboard reducer is modified**. `charts`,
`sliceEntities`, `dataMask`, `dashboardInfo` and `dashboardState` already handle it;
`dashboardLayout` and `nativeFilters` handle it too but dereference their slice without
optional chaining, so the fabricated payload carries an empty stand-in for each.
`datasources` has no hydrate handler at all and is populated through its own action.

**Backend**
- `EmbeddedChart` model and migration (`a1c7e4b62f18`), mirroring `EmbeddedDashboard`
  including `guest_token_revoked_before` and allowed-domain semantics
- `EmbeddedChartDAO` and `GET` / `POST` / `DELETE` on `/api/v1/chart/<pk>/embedded`
- `/embedded/<uuid>` resolves a UUID as either an embedded dashboard or an embedded chart
- `CHART` guest-token resource type, with `has_guest_access_to_chart` matching on the
  embed UUID
- `raise_for_access` previously gated guest datasource access entirely on a `dashboardId`
  in the form data, which a standalone chart never has. Adds a chart leg that authorises
  the datasource only when the guest token was issued for that chart and the request
  targets that chart's own datasource.

**Frontend**
- `src/embedded/embeddedChart/` — fabricated-state hydration, the explore-data hook, and
  the wrapper that renders the dashboard `Chart`
- `Embed chart` in the chart header menu, gated on `can_set_embedded` on `Chart`
- The existing embed modal is reused, with resource-aware copy

**Two fixes that fall out of this**, both of which stand on their own:
- `superset/embedded/view.py` returned a 500 rather than a 403 for a malformed `Referer`
  (for example a host that looks like it carries a non-numeric port). Access already
  failed closed, so this is not a bypass, but any anonymous client could turn one header
  into a stack trace. **This affects embedded dashboards on `master` today** and is
  isolated in its own commit so it can be taken separately.
- `useIsMobile` matches a media query against the current viewport, which inside an iframe
  is the size the host chose for the embed rather than the size of the device. A narrow
  embed on a desktop was served the phone experience and lost its chart controls entirely.
  Gated behind `MOBILE_CONSUMPTION_MODE`, which is off by default.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF

Before: a chart could only be embedded by embedding the dashboard that contained it.

After: charts embedded individually into a third-party page that Superset knows nothing
about — no dashboard behind them, each with its own UUID and guest token, header controls
and interactions intact. Screenshots to follow.

### TESTING INSTRUCTIONS

1. Enable `EMBEDDED_SUPERSET` and set a real `GUEST_TOKEN_JWT_SECRET`.
2. Run the migration: `superset db upgrade` (creates `embedded_charts`).
3. As an Admin, open a dashboard, pick a chart, and choose **Embed chart** from its header
   menu. Set an allowed domain and copy the UUID.
4. Mint a guest token scoped to the chart:
   `POST /api/v1/security/guest_token/` with
   `{"resources": [{"type": "chart", "id": "<uuid>"}], "user": {...}, "rls": []}`
5. From a page served on the allowed domain, embed it with `@superset-ui/embedded-sdk`,
   passing that UUID and a `fetchGuestToken` that returns the token above.
6. Confirm the chart renders with data, and that the header menu still offers Force
   refresh, Enter fullscreen, View query, View as table and Download.

Worth verifying explicitly:
- Requesting `/embedded/<uuid>` with no `Referer`, or one outside the allow-list, returns
  403; the allowed origin returns 200.
- A guest token minted for chart A is refused data for chart B (403), including when both
  charts share a datasource — the check is on the embed UUID, not the dataset.

`hydrateEmbedded.test.ts` runs the real dashboard reducers against the real fabricated
payload, so a slice that a `HYDRATE_DASHBOARD` handler dereferences cannot silently go
missing again; that failure surfaces only at runtime and only in the embedded path.

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [x] Required feature flags: `EMBEDDED_SUPERSET`
- [x] Changes UI
- [x] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [x] Migration is atomic, supports rollback & is backwards-compatible
  - [x] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [x] Introduces new feature or API
- [ ] Removes existing feature or API

The migration only creates a new table and adds no column to an existing one, so it is
additive and reversible. Raising as a draft for early feedback, particularly on the
`HYDRATE_DASHBOARD` reuse and on whether the `raise_for_access` chart leg is scoped
tightly enough.

### CODE REVIEW

File: superset-frontend/src/embedded/embeddedChart/useExploreData.ts [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    data: null,
+    loading: true,
+    error: null,
+  });
+
+  useEffect(() => {
+    let cancelled = false;
+
+    SupersetClient.get({
+      endpoint: `/api/v1/explore/?slice_id=${chartId}`,
```
Comment (gabotorresruiz) [inline superset-frontend/src/embedded/embeddedChart/useExploreData.ts]: This is the load-bearing line for the guest flow: `/api/v1/explore/` sits under `can_read` on Explore, which no embedded guest role has by default. I verified on this branch that a chart guest token gets a 403 here with a guest role that is sufficient for embedded dashboards, so the page dead-ends in the error state. Either the required guest-role grants become a documented part of the feature, or this fetch moves under the Chart permission the guest already needs for `/api/v1/chart/data` (which also means teaching `ChartFilter`'s guest branch about chart tokens, since `GET /api/v1/chart/<pk>` 404s for them today).

File: superset/models/embedded_chart.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    """
+
+    __tablename__ = "embedded_charts"
+
+    uuid = Column(UUIDType(binary=True), default=uuid.uuid4, primary_key=True)
+    allow_domain_list = Column(Text)  # reference the `allowed_domains` property instead
+    # Epoch seconds; guest tokens whose `iat` predates this are rejected. Set to
+    # "now" to revoke all currently-issued guest tokens for this embedded
+    # chart. NULL = no revocation.
+    guest_token_revoked_before = Column(Integer, nullable=True)
```
Comment (gabotorresruiz) [inline superset/models/embedded_chart.py:46]: This column promises revocation semantics that never execute for charts: `_is_guest_token_revoked_by_embedded` skips every non-DASHBOARD resource (`superset/security/manager.py:5339`) and `revoke_guest_token_access` only resolves `EmbeddedDashboardDAO`. I verified it live on this branch: with `guest_token_revoked_before` set in the future on both tables, a dashboard guest token gets a 401 while a chart guest token keeps returning data. Deleting the embed config does cut access, but it destroys the uuid, which is exactly what revocation exists to avoid. I would either add the CHART branch to both spots with a test that sets the cutoff and asserts the 401, or drop the column until charts support it.

File: superset/charts/schemas.py [hunk 3]
```diff
@@ -1998,3 +1998,15 @@ class ChartGetResponseSchema(Schema):
...
+
+
+class EmbeddedChartConfigSchema(Schema):
+    allowed_domains = fields.List(fields.String(), required=True)
+
+
+class EmbeddedChartResponseSchema(Schema):
+    uuid = fields.String()
+    allowed_domains = fields.List(fields.String())
+    chart_id = fields.String()
```
Comment (gabotorresruiz) [inline superset/charts/schemas.py]: Confirming the bots' finding empirically: I created an embed config on a running instance and the response carries no chart identifier at all, because the model exposes `slice_id` and this field never dumps. While fixing it, consider mirroring `EmbeddedDashboardResponseSchema` exactly: the dashboard twin's nested user schema exposes `username`, while this one exposes the `changed_by` user's `email`. Keeping the two response shapes identical fixes both in one move.

File: superset-frontend/src/embedded/index.tsx [hunk 4]
```diff
@@ -47,6 +47,7 @@ import {
   getThemeController,
 } from './EmbeddedContextProviders';
 import { embeddedApi } from './api';
+import EmbeddedChart from './embeddedChart';
```
Comment (gabotorresruiz) [inline superset-frontend/src/embedded/index.tsx]: Not a blocker, but worth doing before merge since CI already measured it: this static import pulls the dashboard chart stack into the initial embedded bundle for every consumer, including existing embedded dashboards that never render a chart embed (+17% on the embedded entrypoint per the benchmark alert). `LazyDashboardPage` at the top of this file is the pattern to copy: a `lazy(() => import('./embeddedChart'))` behind the existing Suspense keeps the chart stack out of the dashboard path.

File: superset/migrations/versions/2026-09-01_10-00_a1c7e4b62f18_add_embedded_charts_table.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+"""add embedded_charts table
+
+Revision ID: a1c7e4b62f18
+Revises: 39097d124752
```
Comment (gabotorresruiz) [inline superset/migrations/versions/2026-09-01_10-00_a1c7e4b62f18_add_embedded_charts_table.py]: Just a small NIT while you re-chain onto master's new head (`b3e9c1a75d24` landed after your last re-chain, which is the current merge conflict): the docstring still says `Revises: 39097d124752` while `down_revision` is `8f31c5d726ab`. For what it is worth, I ran the migration up, down, and up again on this branch and it was clean.

**Review by gabotorresruiz — CHANGES_REQUESTED** [review]
Hi @amaannawab923, thank you for coming back with a genuinely fresh implementation. I stood this branch up locally and the core of my last review is here and working: a chart guest token fetches its own chart's data, is refused another chart's data even on a shared dataset, and is refused arbitrary queries with a spoofed slice_id; bogus mints get a 400; the referrer allow-list, Sec-Fetch-Dest check and postMessage validation all apply to the chart path; the migration cycles cleanly; and your malformed-Referer fix is real (the same request 500s against master's view code and 403s on this branch).

Two things keep me from approving, and they are the same two from last time:

1. The page's first fetch still runs under the Explore permission. With a guest role provisioned the way an embedded-dashboard deployment would be, `GET /api/v1/explore/` returns 403 for a chart guest token and the page dead-ends. It only works after also granting `can_read` on Explore, which the PR neither documents nor tests. Either document the required guest-role grants as part of the feature, or serve the slice/dataset pair under the Chart permission (note `GET /api/v1/chart/<pk>` currently 404s for chart-token guests, so `ChartFilter`'s guest branch would need to learn chart resources for that route). Details inline.
2. The functional tests from the last round are still missing: `has_guest_access_to_chart`'s real body is executed by no test, and the new endpoints, the view resolution, and the CHART validation branch have none. A test minting a chart token and asserting own-chart 200, cross-chart 403, and bogus-uuid 400 would pin everything I verified by hand.

One new find inline: chart guest tokens currently ignore `guest_token_revoked_before` entirely. Smaller items also inline. This is much closer than #33424 ever was, and the data-layer security model is right; with the metadata path settled and the tests in, I would be glad to approve.

File: superset/charts/schemas.py [hunk 6]
```diff
@@ -1998,3 +1998,15 @@ class ChartGetResponseSchema(Schema):
...
+    changed_on = fields.DateTime()
+    changed_by = fields.Nested(UserSchema)
```
[the remark this answers is not in the document]
Comment (amaannawab923) [reply superset/charts/schemas.py]: Fixed. `chart_id` is now `fields.String(attribute="slice_id")`, so it dumps from the column the model actually has rather than resolving to nothing. Verified against a running instance: the embedded-chart response now carries `"chart_id": "103"`.

File: superset/charts/api.py [hunk 7]
```diff
@@ -1874,3 +1897,181 @@ def restore_version(self, uuid_str: str, version_uuid_str: str) -> Response:
...
+              $ref: '#/components/responses/500'
+        """
+        chart = ChartDAO.find_by_id(pk)
+        if not chart:
+            return self.response_404()
+        try:
+            body = self.embedded_config_schema.load(request.json)
+            embedded = EmbeddedChartDAO.upsert(chart, body["allowed_domains"])
+            db.session.commit()  # pylint: disable=consider-using-transaction
+            result = self.embedded_response_schema.dump(embedded)
```
[the remark this answers is not in the document]
Comment (amaannawab923) [reply superset/charts/api.py:2261]: Fixed in the schema rather than here — `chart_id` is now bound to the model's `slice_id` attribute, so this endpoint's response carries the identifier. Confirmed on a running instance.

File: superset/daos/chart.py [hunk 8]
```diff
@@ -166,3 +167,33 @@ def remove_favorite(chart: Slice) -> None:
...
+
+    @staticmethod
+    def upsert(chart: Slice, allowed_domains: list[str]) -> EmbeddedChart:
+        """
+        Sets up a chart to be embeddable.
+        Upsert is used to preserve the embedded_chart uuid across updates.
+        """
+        embedded: EmbeddedChart = (
+            chart.embedded[0] if chart.embedded else EmbeddedChart()
+        )
```
[the remark this answers is not in the document]
Comment (amaannawab923) [reply superset/daos/chart.py:184]: Looked at this properly and I would rather not fix it here, though I want to lay out why since the observation is correct.

`EmbeddedDashboardDAO.upsert` has the identical read-then-write against its own relationship, and neither `embedded_dashboards.dashboard_id` nor `embedded_charts.slice_id` carries a unique constraint in any migration. I did consider adding one to `embedded_charts` while the migration is still unshipped, but both models document the current behaviour as deliberate — "This data model allows multiple configurations for a given [dashboard/chart], but at this time the API only allows setting one" — so a constraint would contradict the model's own stated contract, and only on the new table.

If that contract should tighten to one configuration per resource, the change covers both tables and both DAOs and is worth doing on its own rather than folded into this one. Happy to raise it separately.

File: superset-frontend/src/dashboard/components/EmbeddedModal/index.tsx [hunk 9]
```diff
@@ -59,15 +62,19 @@ const ButtonRow = styled.div`
...
+}: Props) => {
   const { addInfoToast, addDangerToast } = useToasts();
   const [ready, setReady] = useState(true); // whether we have initialized yet
   const [loading, setLoading] = useState(false); // whether we are currently doing an async thing
   const [embedded, setEmbedded] = useState<EmbeddedDashboard | null>(null); // the embedded dashboard config
   const [allowedDomains, setAllowedDomains] = useState<string>('');
   const [showDeactivateConfirm, setShowDeactivateConfirm] = useState(false);
 
-  const endpoint = `/api/v1/dashboard/${dashboardId}/embedded`;
+  const endpoint = `/api/v1/${resourceType}/${dashboardId}/embedded`;
```
[the remark this answers is not in the document]
Comment (amaannawab923) [reply superset-frontend/src/dashboard/components/EmbeddedModal/index.tsx:77]: Fixed. The effect depended on `[dashboardId]` while the `endpoint` it calls is derived from `resourceType` as well, so it now depends on `[endpoint]`, which covers both.

Worth noting for accuracy: both call sites pass a fixed `resourceType` for the component's lifetime today (`SliceHeaderControls` always `chart`, `Header` always the default), so this is a latent stale-closure hazard rather than a reachable bug. The dependency was wrong either way. Added a test that flips `resourceType` via `rerender` while holding `dashboardId` constant.

File: superset-frontend/src/embedded/embeddedChart/hydrateEmbedded.ts [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+        slices: {
+          [key]: {
+            slice_id: key,
+            slice_url: slice.slice_url,
+            slice_name: slice.slice_name,
+            form_data: slice.form_data,
+            viz_type: slice.form_data.viz_type,
+            datasource: slice.form_data.datasource,
+            description: slice.description,
+            description_markeddown: slice.description_markeddown,
```
[the remark this answers is not in the document]
Comment (amaannawab923) [reply superset-frontend/src/embedded/embeddedChart/hydrateEmbedded.ts]: Fixed, though the diagnosis is slightly different from the one here. `description_markeddown` is the API's own spelling — it is the field name on the chart schema, and both `hydrate.ts` and `sliceEntities.ts` read it and write the value out under the correctly spelled `description_markdown` key that `Chart.tsx` reads.

The bug was that this payload passed the API spelling straight through instead of performing that rename, so the chart stack never found the field. The input name is unchanged; the output key is now `description_markdown`, matching the two existing handlers. Added a regression test.

File: superset-frontend/src/embedded/embeddedChart/useExploreData.ts (hunk 1 above)
Comment (amaannawab923) [reply superset-frontend/src/embedded/embeddedChart/useExploreData.ts]: Moved the fetch rather than documenting the grant, and you were right that it needed the `ChartFilter` work too.

The reason I did not go the documentation route: `/api/v1/explore/` builds its payload from request-supplied `form_data` and resolves the datasource it returns from that rather than from the `slice_id`, so a grant there reaches well past the one chart a token names — and since `GUEST_ROLE_NAME` defaults to `Public`, that grant would not have been scoped to embedding either.

The embed now reads `GET /api/v1/chart/<pk>/embedded_context`, which returns the chart and its dataset in one payload, sits under `Chart` read, and resolves one fixed chart. It is the chart analogue of a dashboard's `/charts` and `/datasets` sub-resources, collapsed into one call since a chart has exactly one of each.

`ChartFilter`'s guest branch was worse than the 404 you saw: a chart-only token produced an empty dashboard scope, which the branch turned into a deny-all clause, so *every* Chart-API read was refused for exactly the tokens this feature issues. It now unions the token's dashboard scope with an `EXISTS` over the chart's own embed rows, and a token granting neither still denies everything.

Checked on a running instance with the guest role holding only the shipped `PUBLIC_ROLE_PERMISSIONS` (no Explore grant): `/api/v1/explore/?slice_id=103` returns 403, `/api/v1/chart/103/embedded_context` returns 200, and a token for one chart gets 404 on another. Three charts render end to end through the SDK on a third-party page.

One thing worth flagging since it differs from the dashboard path: for the single dataset the caller was just authorised on, a guest holding that chart's token keeps the rendering metadata — `columns`, `metrics`, `verbose_map` — which the dashboard's rule strips. Without it the chart cannot draw. `params` stays withheld even then, being operator-authored configuration the renderer never reads, and `DashboardDatasetSchema`'s own `post_dump` still removes connection and query internals for any guest.

File: superset/models/embedded_chart.py (hunk 2 above)
Comment (amaannawab923) [reply superset/models/embedded_chart.py:46]: Added the CHART branch to both spots and kept the column.

`_is_guest_token_revoked_by_embedded` now dispatches on resource type instead of skipping everything that is not a dashboard, and `revoke_guest_token_access` retries the uuid against `EmbeddedChartDAO` when it does not resolve to an embedded dashboard. The two uuid spaces are distinct, so the fallthrough is unambiguous — the same ordering `EmbeddedView.embedded` already uses. Charts are only ever addressed by their embed uuid, so there is no equivalent of the dashboard branch's legacy raw-id lookup.

Tests mirror the existing dashboard cases: cutoff in the future rejects, a token issued after the cutoff is accepted, no cutoff accepts, a token with no `iat` is rejected when a cutoff is set, an unresolvable uuid leaves nothing to enforce, plus a mixed dashboard-and-chart token revoked by the chart's cutoff alone.

One difference from what you observed: on this endpoint a revoked chart token comes back 404 rather than 401, because the guest never resolves and `ChartFilter` then excludes the row. Same outcome, different route. Confirmed on a running instance — the same token returned 200 before the cutoff and 404 after.

The request-loader test needed one precaution worth mentioning: `get_guest_user_from_request` swallows every failure into `None`, so asserting only "returns `None`" would pass for unrelated reasons. It replays the identical request twice, changing only the cutoff.

File: superset/charts/schemas.py (hunk 3 above)
Comment (amaannawab923) [reply superset/charts/schemas.py]: Fixed, and thanks for pinning it down empirically — that made it unambiguous.

`chart_id` is now `fields.String(attribute="slice_id")`, so it dumps from the column the model actually has. I kept the public field name rather than renaming to `slice_id`: the convention across both embedded response schemas is one field named after the parent resource type, and the endpoint's own docs already describe the path parameter as the chart id.

Took the mirroring suggestion too. `changed_by` now points at a schema dedicated to this response that matches `EmbeddedDashboardResponseSchema`'s nested user schema field for field, rather than changing the shared `UserSchema` that other chart endpoints dump through.

On a running instance the two responses are now the same shape:

```
chart:      {"uuid": "...", "allowed_domains": [...], "chart_id": "103",   "changed_on": "...", "changed_by": null}
dashboard:  {"uuid": "...", "allowed_domains": [...], "dashboard_id": "9", "changed_on": "...", "changed_by": null}
```

File: superset-frontend/src/embedded/index.tsx (hunk 4 above)
Comment (amaannawab923) [reply superset-frontend/src/embedded/index.tsx]: Done, following `LazyDashboardPage` exactly — `lazy(() => import('./embeddedChart'))` with its own `webpackChunkName`. It renders inside the `Suspense` boundary that component already sits in, so no new boundary was needed.

The split is real rather than just structural: the build now emits `EmbeddedChart.<hash>.chunk.js` separately, and in a dev build that chunk is 3.2M against a 384K `embedded` entry. Dashboard-only consumers no longer pay for any of it.

File: superset/migrations/versions/2026-09-01_10-00_a1c7e4b62f18_add_embedded_charts_table.py (hunk 5 above)
Comment (amaannawab923) [reply superset/migrations/versions/2026-09-01_10-00_a1c7e4b62f18_add_embedded_charts_table.py]: Both fixed. The docstring now matches, and the chain points at `7e2c9a4f1b83`.

The mismatch turned out to be the more interesting half: the migration and `b3e9c1a75d24` both claimed `8f31c5d726ab` as their parent, which is what left two heads and produced the conflict. Thanks for running it up and down — that saved a round trip.

File: superset/charts/api.py [hunk 11]
```diff
@@ -2016,3 +2040,272 @@ def restore_version(self, uuid_str: str, version_uuid_str: str) -> Response:
...
+        # is exactly one and it belongs to the chart the caller was just
+        # authorized on, so a guest holding a token for that chart keeps the
+        # rendering metadata -- columns, metrics, verbose map -- it needs to draw
+        # the chart. ``params`` is withheld even then: it is operator-authored
+        # free-form configuration rather than anything the renderer reads.
+        entitled_guest = security_manager.has_guest_access_to_chart(chart)
+        if not (security_manager.can_access_datasource(datasource) or entitled_guest):
+            for key in DASHBOARD_DATASET_INACCESSIBLE_FIELDS:
+                dataset.pop(key, None)
+        elif entitled_guest:
```
Comment (gabotorresruiz) [inline superset/charts/api.py:2170]: This block worries me a bit. For a guest holding this chart's token only `params` is dropped, so `columns`, `metrics` and `verbose_map` all go out with the response. On my instance that guest received every column of the dataset, including one the chart never references, plus each metric's SQL `expression`.

The same guest on the dashboard path receives none of those three: `_serialize_dashboard_dataset` (`superset/dashboards/api.py:759`) drops them for any caller without datasource access, which is what #42716 put them into `DASHBOARD_DATASET_INACCESSIBLE_FIELDS` for.

Are we certain the renderer needs them? An embedded dashboard draws the same `Chart` component from the same `datasources` slice (`DashboardPage.tsx:379`, fed by `useDashboardDatasets`), and for a guest that payload already arrives without those fields. If something does break without them, narrowing `columns` and `metrics` to what the chart's own `form_data` references would keep the chart drawable without publishing the rest of the dataset alongside it. Either way, a test asserting the guest payload carries no column outside the chart's `form_data` and no metric `expression` would pin whichever way you land.

**Review by gabotorresruiz — CHANGES_REQUESTED** [review]
Thanks for the rework @amaannawab923. I rebuilt the branch and exercised it live, and both blockers from my last review are genuinely resolved.

- The metadata no longer loads under Explore. On a fresh instance with the shipped guest role (`Public`, which carries `can read on Chart` and no Explore grant), a chart guest token gets 403 from `GET /api/v1/explore/?slice_id=1` and 200 from `GET /api/v1/chart/1/embedded_context`. The scope stays tight around it: the other chart on the same dataset is 404 there and 403 on `/api/v1/chart/data`, `GET /api/v1/chart/` returns exactly one chart, the dashboard endpoints return 404, and an ad hoc metric over a column the chart does not use is refused.
- Revocation now covers charts. With a cutoff written to the `embedded_charts` row the same token goes from 200 to 403 on `/api/v1/chart/data` and to 404 on `embedded_context`, then back to 200 once I clear it.

Also confirmed on the running branch: guest RLS applies to a chart token (the same query returns 1 row with a `region = 'east'` rule and 2 rows without), the referrer allow-list answers 200 for an allowed origin and 403 for disallowed, malformed and missing ones, `chart_id` dumps as `"1"` mirroring the dashboard response field for field, and the migration goes up, down and up cleanly on SQLite with `embedded_charts` matching the model.

Two things left, one of them inline.

#### The endpoint test from last round

`has_guest_access_to_chart` is the single function deciding whether a token grants a chart, and no test runs its real body (`tests/unit_tests/security/test_embedded_guest_chart_access.py:66` and `:103` mock it), while none of the four new endpoints has a functional test of its own.

The filter, revocation and schema tests you added are solid, and I checked they genuinely pin the fixes: neutralising the chart branches in `superset/security/manager.py` and `superset/charts/filters.py` fails 7 of them. What is still missing is the one from last round: mint a chart token, then assert own chart 200, cross chart 403 and a bogus uuid 400. That is exactly the surface I had to re-verify by hand again this time.

Happy to approve once that lands and the dataset question below is settled either way.
